### ISSUE TITLE
Cross Validation and custom pipeline in Autogluon.

### ISSUE THREAD

**Issue by salad-henchmen** [awslabs/autogluon#999]
Hello,
I have a problem with data leakage and wanted to avoid it so I created a custom pipeline to cleanse the data before train/validation. However, when the cross validation is performed across the train dataset, splitting it into train and val in each fold, does the pair go through the _fit_transform within the pipeline? Is there any extra configuration that I'll have to enable first?

**Comment by Innixma** [comment]
Could you provide more context in terms of the form of data leakage that is in your dataset and how it relates to the cross-validation splits?

**Comment by salad-henchmen** [comment]
Hello! Sorry for the late response. I'm performing an aggregate operation across an entire feature. And so I'm trying to build a separate pipeline to handle the validation data such that this operation is performed across this val_data using the information we obtain from the train set. I was wondering how cross validation is performed so that for each temporary train and validation set, i could transform them separately. 
Thank you!

**Comment by Innixma** [comment]
Ah I see! This does sound like useful functionality. We don't currently support this for the global preprocessing step, however for the model-specific preprocessing, one option is to create custom models and specify them in hyperparameters. These custom models could override the `_preprocess` method by first calling your custom feature engineering logic that is split specific. (data passed to `_preprocess` is fit with only the train data for a particular model, so there wouldn't be data leakage)

This is a bit complicated and isn't the ideal solution, since you'd need to do this for every model you intend to use. I've added this request to our roadmap.

In the meantime, the suggested option would be to handle as much feature engineering as possible outside of AutoGluon and then use `tuning_data` and avoid bagging/stacking.

Also, if this data happens to be public, we'd love to get examples of such data to add to our benchmarking. This would help significantly in our prioritization of the feature.

**Comment by salad-henchmen** [comment]
Oh okay! I'll look into that too then. But how about using the AbstractFeatureGenerator to build a custom pipeline. Can't the _fit_transform and _transform be used to handle the train and validation datasets separately? Also, wouldn't this allow me to continue using auto_stack safely?
